src/core/search/block_list.cc

```diff
@@ -0,0 +1,128 @@
+#include "core/search/block_list.h"
+
+namespace dfly::search {
+
+using namespace std;
+
+template <typename C> bool BlockList<C>::Insert(DocId t) {
+  auto block = FindBlock(t);
+  if (block == blocks_.end())
+    block = blocks_.insert(blocks_.end(), C{blocks_.get_allocator().resource()});
+
+  if (!block->Insert(t))
+    return false;
+
+  size_++;
+  TrySplit(block);
+  return true;
+}
+
+template <typename C> bool BlockList<C>::Remove(DocId t) {
+  if (auto block = FindBlock(t); block != blocks_.end() && block->Remove(t)) {
+    size_--;
+    TryMerge(block);
+    return true;
+  }
+
+  return false;
+}
+
+template <typename C> typename BlockList<C>::BlockIt BlockList<C>::FindBlock(DocId t) {
+  DCHECK(blocks_.empty() || blocks_.back().Size() > 0u);
+
+  if (!blocks_.empty() && t >= *blocks_.back().begin())
+    return --blocks_.end();
+
+  // Find first block that can't contain t
+  auto it = std::upper_bound(blocks_.begin(), blocks_.end(), t,
+                             [](DocId t, const C& l) { return *l.begin() > t; });
+
+  // Move to previous if possible
+  if (it != blocks_.begin() && blocks_.size() > 0)
+    --it;
+
+  return it;
+}
+
+template <typename C> void BlockList<C>::TryMerge(BlockIt block) {
+  if (block->Size() == 0) {
+    blocks_.erase(block);
+    return;
+  }
+
+  if (block->Size() >= block_size_ / 2 || block == blocks_.begin())
+    return;
+
+  // Merge strictly right with left to benefit from tail insert optimizations
+  size_t idx = std::distance(blocks_.begin(), block);
+  blocks_[idx - 1].Merge(std::move(*block));
+  blocks_.erase(block);
+
+  TrySplit(blocks_.begin() + (idx - 1));  // to not overgrow it
+}
+
+template <typename C> void BlockList<C>::TrySplit(BlockIt block) {
+  if (block->Size() < block_size_ * 2)
+    return;
+
+  auto [left, right] = std::move(*block).Split();
+
+  *block = std::move(right);
+  blocks_.insert(block, std::move(left));
+}
+
+template <typename C>
+typename BlockList<C>::BlockListIterator& BlockList<C>::BlockListIterator::operator++() {
+  ++*block_it;
+  if (block_it == block_end) {
+    ++it;
+    if (it != it_end) {
+      block_it = it->begin();
+      block_end = it->end();
+    } else {
+      block_it = std::nullopt;
+      block_end = std::nullopt;
+    }
+  }
+  return *this;
+}
+
+template class BlockList<CompressedSortedSet>;
+template class BlockList<SortedVector>;
+
+bool SortedVector::Insert(DocId t) {
+  if (entries_.size() > 0 && t > entries_.back()) {
+    entries_.push_back(t);
+    return true;
+  }
+
+  auto it = std::lower_bound(entries_.begin(), entries_.end(), t);
```

Comment:
binary_search?

Reply:
lower bound is binary search, the `binary_search` you're referring to is something else

Reply:
lowerv_bound finds the first element in the range [first, last) such that element < value. But you do check next line if (it != entries_.end() && *it == t)
and it looks like you need an exact element that binary search do

Reply:
I need the iterator, binary search doesn't give me one, using lower bound is standard practice for finding elements